### DISCUSSION TITLE
Pre-released SpamBlocker for testing the API

### DISCUSSION THREAD

**Discussion by aj3423** [haumacher/phoneblock#115] [Show and tell]
The query/report features are done, I just released it for public testing.
Only the testing api is enabled: https://phoneblock.net/pb-test/api,  I'll switch to the production api later in the official release.

I limited the user input like
- The username must be a uuid4 format
- The password must match base64(16 bytes)
You won't change this often, right? :)

Please let me know if you notice anything wrong in the database.

**Comment by haumacher** [comment]
Great, how can I easily install this test version?

**Comment by aj3423** [comment]
From the latest release: https://github.com/aj3423/SpamBlocker/releases/tag/latest
These two are not implemented yet and will be available in the final release.
- Manual report 
- Report a number blocked by API back to the API, even for those blocked in spam db.

**Comment by haumacher** [comment]
Thanks, I installed the test version. However, I cannot confirm that it works. I enabled the blocker and installed a user name for the request and report APIs (all other options are turned off). Then I called my phone from my landline. The phone rings, but SpamBlocker does not show the calling number in its calls tab. Is there something else I have to enable?

I would have expected that the call is listed and I should see a request in PhoneBlock's logs.

**Reply by aj3423** [reply]
You need to turn this on 
![Screenshot_20241208-170259_SpamBlocker.png](https://github.com/user-attachments/assets/67c9c460-2d9e-47eb-a319-dc2911f4afbc)

And when a call is blocked, it will report after an hour, you can see that reporting log in the  logcat.

if you want to manually test, you can use the test button, It will report immediately
![Screenshot_20241208-170711_SpamBlocker.png](https://github.com/user-attachments/assets/e40c482a-1c54-4e95-88f5-23586f39690f)

**Comment by haumacher** [comment]
I had the following settings:
<img src="https://github.com/user-attachments/assets/63fb00cb-1a95-4b4e-a786-7d2f2bedf093" height="500px">

<img src="https://github.com/user-attachments/assets/cdc28569-5071-4fee-b2af-1e763c19e698" height="500px">

But even after calling my phone, the calls tab stays empty:

<img src="https://github.com/user-attachments/assets/2316b7a8-a175-4d19-b609-609e23df4c8b" height="500px">

**Reply by aj3423** [reply]
Have you set this app as the caller ID app? When you tap on the call tab it will prompt this, have you rejected it?

**Reply by haumacher** [reply]
No did accept  this.

**Reply by haumacher** [reply]
Can I verify this afterwards in some way?

**Reply by haumacher** [reply]
<img src="https://github.com/user-attachments/assets/ea167e98-f339-4eab-9ac6-d76f66d93406" height="500px">

See the first entry.

**Reply by haumacher** [reply]
And yes, I've enabled "show accepted" and "show blocked" in the calls tab.

**Reply by aj3423** [reply]
The call you made previously, was that a contact number?

**Reply by aj3423** [reply]
That's really strange, are you running in work profile ?

**Reply by haumacher** [reply]
Yes the caller (my landline) is in my contacts, but I did not enable the rule for processing contacts separately.

What do you mean with "work profile"?

**Reply by aj3423** [reply]
I see the reason, because it's a contact number, and the app hasn't granted the contact permission, so Android will just let a  call ring and the call screening services not triggered

**Reply by aj3423** [reply]
please turn the Contacts switch on, then it should work
![Screenshot_20241208-173647_SpamBlocker.png](https://github.com/user-attachments/assets/0fcb90d0-2a1e-4b5d-a529-0b73e560c481)

**Reply by haumacher** [reply]
Yes this worked - I enabled the contacts rule - gave permission and disabled the rule afterwards. Then the call is shown.

**Reply by haumacher** [reply]
And yes, the number was checked by PhoneBlock, and yes, if I block my number in the PhoneBlock DB, I no longer can call my phone, great!

**Reply by aj3423** [reply]
Great!

**Reply by aj3423** [reply]
This number is blocked by the API and currently is will not be reported. To test the report, you can add a regex rule `.*` , then non-contact calls will be blocked by this rule and will be scheduled to report. 

You can simulate a call with this button 
![Screenshot_20241208-175043_SpamBlocker.png](https://github.com/user-attachments/assets/7c5b8d77-1c97-4969-89b6-a31b3001c489)

**Reply by aj3423** [reply]
I'll be off for a while:)

**Comment by haumacher** [comment]
Don't see the reporting feature yet. When I long click the accepted call I get the options 

* copy number
* add to rule
* add to db
* mark read

how to report?

**Reply by haumacher** [reply]
What I mean here is the explicit reporting feature. I long clicked the first call that initially was accepted. I was expecting that I now can manually report the number to PhoneBlock.

**Reply by aj3423** [reply]
currently, it only supports auto reporting. The manual reporting requires changes to the UI, the UI is not my thing, I'm trying to learn how to implement a swipe menu, it will take a couple of days.
if you want to try manual reporting, you can tap the item in the section Report API, and tap the Test button, input a phone number, It will report immediately.

**Reply by haumacher** [reply]
OK, got it.

**Comment by haumacher** [comment]
I've some other suggestions regarding the UI of SpamBlocker. The functionality is really great for expert users (defining custom rules/service lookups/workflows). However, the average user is already extremely challenged when he should copy three values (URL, user name, password) from the PhoneBlock website to the Fritz!Box configuration interface (25% fail during this step).  

To make SpamBlocker usable after installation (with PhoneBlock), the user has to
     1.  navigate to settings,
     2. click the enable button
     3. find the new button for "request API", 
     4. select PhoneBlock, 
     5. open the PhoneBlock website,
     6. register there
     7. copy the user name
     8. copy the password
     9. repeat the same with the report API,
     10. during all steps he must not click on the wrong part of the UI - e.g. when clicking on the request configuration, it is very easy to drag and drop a step to the wrong location. It happened to myself and it was hard to fix this afterwards.  

To enable an average user to get rid of SPAM calls using your app, you should consider the following:
   1. Automatically launch a setup wizard during first startup. This should ask the minimum amount of questions/provide a reasonable choice of default options to make the app usable, e.g: 
       * Enable STIR check (but give it a comprehensible name)
       * Add FTC database
       * Add PhoneBlock
   2. Split the settings into normal and expert ones, I would consider the following normal (where you cannot easily break things):
      * block non-contacts
      * block repeated calls
      * white-list called numbers
      * block type
   3. Provide an integrated setup for services such as PhoneBlock, I suggest the following workflow
      * When the user clicks "enable PhoneBlock", open a webview with the PhoneBlock website with a yet to specify URL
      * The displayed page should allow the user either to log-in or to create an account. 
      * After the login is finished, the user is redirected to a setup page that communicates an authorization token to the SpamBlocker app using a JavascriptChannel. 
      * This authorization token is then stored by SpamBlocker internally in secure storage and used for authentication of either request API or report API.

What do you think?

**Reply by aj3423** [reply]
Thanks for learning the app, and for these suggestions, but I have no plan for them.

1. I dislike wizards,  the only wizard I'd consider is one that guides users to read balloon tooltips, which are constantly being refined.
2. About the "normal" and "expert" settings, the quick-settings are the "normal" ones, the regex filter requires users to have a higher level of expertise, and even higher level when it goes down. 
I've thought about hiding some sections by default and add a button "show advanced features", but that's not what you meant.
3. Webview is too much work, there can also be other issues like website compatibility, I think a clickable link is enough.

I'll be focusing on blocking calls, for those who really wants to spend time on learning the app. I anticipate a 80% of user loss after they tried it, most of them would want a one-click solution, and I won't spend 5 days implementing any feature for those who don't want to spend 5 minutes reading the tooltips.

**Reply by haumacher** [reply]
Thanks for your clear words, I really appreciate that!

**Comment by aj3423** [comment]
Hi, I got some questions.
When manually reporting, it's supposed to report the number to all API endpoints, different endpoints may have different categories. 

For example, PhoneBlock has the "gamble", others might have "stock", or "cryptocurrency", there would be too many category buttons on the UI. So, instead of supporting all of them, I decide to only provide some generic categories, like:

![image](https://github.com/user-attachments/assets/c2954f1e-9267-4b87-a903-1f17b01a9978)

For PhoneBlock:
- "Valid Number" corresponds to -> A_LEGITIMATE 
- "Other Spam" -> B_MISSED
- "Survey" -> D_POLL
- "Marketing" -> E_ADVERTISING
- "Fraud" -> G_FRAUD

(I guess the C_PING is for testing purpose, it's ignored.)

And the problems are:
1. It will never get reported as F_GAMBLE
  In my view, promoting gambling is essentially marketing, classifying it as a standalone category would be overly specific in this app.

2. Political 
  PhoneBlock doesn't have this category, I'd suggest to use the "B_MISSED" as the fallback type, when user reporting it as "Political", it will be reported as "B_MISSED". I'm not sure if there are any more categories to add in the future, but if there are, they can also be reported as the fallback category.

Are you okay with that? Or any idea?

**Reply by haumacher** [reply]
It is OK to never report "gamble", it is either marketing or more often fraud.

Political - you mean marketing for some political party? I had never such call in Germany, but I suggest to better also use "marketing" for that.

**Reply by aj3423** [reply]
> Political - you mean marketing for some political party?

Usually they promote a specific candidate or party. I'll use the "marketing" for it, thanks.
